### PR TITLE
Fixed #35742 -- Removed hardcoded "username" references in admin templates.

### PR DESCRIPTION
**Pull request by Antoliny0919**
#### Trac ticket number

[ticket-35742](https://code.djangoproject.com/ticket/35742#ticket)

#### Branch description
Improve UserAdmin by replacing the static "username" with the dynamic USERNAME_FIELD value.
```
# admin.py
class CustomUserAdmin(UserAdmin):
  add_fieldsets = (
    (
      None,
      {
        "classes": ("wide",),
        "fields": (User.USERNAME_FIELD, "usable_password", "password1", "password2"),
      },
    ),
  )

# add_form.html
{% block form_top %}
  {% if not is_popup %}
    <p>{% blocktranslate %}First, enter a {{ username }} and password. Then, you’ll be able to edit more user options.{% endblocktranslate %}</p>
  {% else %}
    <p>{% blcoktranslate %}Enter a {{ username }} and password.{% endblocktranslate %}</p>
  {% endif %}
{% endblock %}
```

#### Checklist
- [x] This PR targets the `main` branch. 
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

File: django/contrib/auth/admin.py
Comment (cliff688) [inline django/contrib/auth/admin.py:1]: I'm not sure about the helper function. I think if we have better ways of passing ``USERNAME_FIELD.verbose_name`` into the template. Maybe we can add it into ``extra_context`` in [add_view()](https://github.com/django/django/blob/826ef006681eae1e9b4bd0e4f18fa13713025cba/django/contrib/auth/admin.py#L119C1-L152C66) after line 151
```diff
    extra_context['username_field_name'] = username_field.verbose_name
```
and then in the template we would then use ``username_field_name`` which would be in the context.

If so then we may have to rewrite the sentence so that we do away with the indefinite particles, as I see no way of dealing with them: "a username" might become "a email" which would be grammatically incorrect.

**Review by cliff688 — COMMENTED** [review]
Thank you for starting this @Antoliny0919 💪 !
I think this might take a few iterations before we end up with something we're happy with.
I have added a comment, please let me know what you think.

File: django/contrib/auth/admin.py
Comment (Antoliny0919) [reply django/contrib/auth/admin.py:1]: Thank you so much for the advice! @cliff688 
As Cliff688 said, I decided that it would be better to add the process in the _add_view method than to add username_field data to the context by overwriting the overridden method render_change_form method.
```
    def _add_view(self, request, form_url="", extra_context=None):
        ...
        defaults = {
            "auto_populated_fields": (),
            "username_help_text": username_field.help_text,
            "username_field_name": username_field.verbose_name,
        }
        ...
```
Thank you so much for your advice!!

**Review by cliff688 — COMMENTED** [review]
Hi @Antoliny0919 

I've been looking at admin and auth_tests and I'm not sure where to add the tests for this. Do you have any ideas?

**Comment by cliff688** [comment]
We may also have to casefold the username_field_name.

**Comment by Antoliny0919** [comment]
@cliff688 
Thank you for your advice on casefold.
I think so too, casefold is also a part that needs to be added.
```
    def _add_view(self, request, form_url="", extra_context=None):
        ...
        defaults = {
            "auto_populated_fields": (),
            "username_help_text": username_field.help_text,
            "username_field_name": username_field.verbose_name.casefold(),
        }
        ...
```
[How about adding a test code to this file?](https://github.com/django/django/blob/main/tests/auth_tests/test_templates.py) --> (tests/auth_tests/test_templates.py)
Should I add a test code for changes to the method in add_view as well?

**Comment by Antoliny0919** [comment]
I found a good example.
On the Django Admin login page, the message that appears when the login fails uses USERNAME_FIELD.
```
class AuthenticationForm(forms.Form):
    def __init__(self, request=None, *args, **kwargs):
        ...
        self.username_field = UserModel._meta.get_field(UserModel.USERNAME_FIELD)

    def get_invalid_login_error(self):
        return ValidationError(
            self.error_messages["invalid_login"],
            code="invalid_login",
            params={"username": self.username_field.verbose_name},
        )
```
The params dictionary username value delivered when an error occurs is the same as the username_name_field I added to the _add_view method. (Access the verbose_name attribute to the same object.)

And I found one more problem.
Screenshot 2024-09-11 at 8 39 32 PM

I confirmed that the translate is not applied.
I think, I need to modify the pofile of all languages for this part.

**django/contrib/admin/locale/all/LC_MESSAGES/django.po**
```
...
msgid ""
"First, enter a %(username_name_field)s and password. Then, you’ll be able to edit more user "
"options."
msgstr ""
"첫 번째로, %(username_name_field)s과 비밀번호를 입력하세요. 그 후, 독자는 더 많은 사용자 옵"
"션을 수정할 수 있습니다. "
...
```
How about modifying username_field_name to username for consistency?
It doesn't seem to cause problems even if it's used as a username key.
```
def _add_view(self, request, form_url="", extra_context=None):
        defaults = {
            "auto_populated_fields": (),
            "username_help_text": username_field.help_text,
            # "username_field_name": username_field.verbose_name,
            "username": username_field.verbose_name,
        }
```

**Comment by cliff688** [comment]
> @cliff688 Thank you for your advice on casefold. I think so too, casefold is also a part that needs to be added.

Actually I found out [in the docs](https://docs.djangoproject.com/en/5.1/topics/db/models/#verbose-field-names) that

> The convention is not to capitalize the first letter of the [verbose_name](https://docs.djangoproject.com/en/5.1/ref/models/fields/#django.db.models.Field.verbose_name). Django will automatically capitalize the first letter where it needs to.

So we don't have to casefold the name

**Comment by cliff688** [comment]
> How about modifying username_field_name to username for consistency?
> It doesn't seem to cause problems even if it's used as a username key.

I don't see any problem with that.

**Comment by Antoliny0919** [comment]
I have updated all the .po files so that the username template variable can be translated.
Only the part that was previously used as 'username' was changed to %(username)s, but some languages were unable to change the 'username' part because it was difficult to distinguish exactly.
What should I do about these parts?

**Comment by Antoliny0919** [comment]
I have updated the release for the changes.
```
:mod:`django.contrib.admin`
~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
* The ``admin/auth/user/add_form.html`` input guideline message is applied
  with a value according to user model username_field.
...

:mod:`django.contrib.auth`
~~~~~~~~~~~~~~~~~~~~~~~~~~
...
* :meth:`.UserAdmin._add_view` The username_field verbose_name attribute of the
  user model was added as a username key to the defaults dictionary.
...
```
I don't know if the release was written well. I need a feedback. ✋

File: django/contrib/admin/templates/admin/auth/user/add_form.html [hunk 1]
```diff
@@ -3,9 +3,9 @@
 
 {% block form_top %}
   {% if not is_popup %}
-    <p>{% translate 'First, enter a username and password. Then, you’ll be able to edit more user options.' %}</p>
+    <p>{% blocktranslate %}First, enter a {{ username }} and password. Then, you’ll be able to edit more user options.{% endblocktranslate %}</p>
```
Comment (claudep) [inline django/contrib/admin/templates/admin/auth/user/add_form.html:6]: Unfortunately, you cannot simply dump a verbose_name inside a sentence. If it works in English, it will fail in many languages where the word is spelled differently depending its position in the sentence. Even in English, it may fail for example if the username field is `identifier`, where the article should then be `an`. We may workaround this by some construct like: `First, fill the fields: {{ username }} and password. Then…`.

I reckon the `AuthenticationForm.error_messages` already presents such a translatable string, so it should probably be rephrased too. We should at least add a [translator comment](https://docs.djangoproject.com/en/stable/topics/i18n/translation/#comments-for-translators) to explain the issue to translators.

Comment (Antoliny0919) [reply django/contrib/admin/templates/admin/auth/user/add_form.html:6]: Thank you so much for your review. @claudep 
I also checked the answers you wrote on the ticket.

Is it correct to understand that the characteristics of each language are inappropriate to express using 'username' as a template variable in the sentence 'First, enter ausername, and password'?
```
<p>{% blocktranslate %}First, fill the fields: {{ username }} and password. Then, you’ll be able to edit more user options.{% endblocktranslate %}</p>
```
I don't need to modify the relate .po file when I change it as above, right?

Comment (claudep) [reply django/contrib/admin/templates/admin/auth/user/add_form.html:6]: Yes, it is correct. Each language has its own rules regarding the spelling of words depending on its grammatical position in a sentence.

Yes, avoid modifying .po files, that's the translator's job.

**Comment by nessita** [comment]
@Antoliny0919 Thank you for continuing the work in this PR. :star2: 

I will update the ticket and I will set the "patch needs tests" and "patch needs improvement" flag to signal that there is still a need for:
1. fix the docs building errors, and
2. add tests for this change.

Looking forward to those fixes!

**Comment by Antoliny0919** [comment]
I have created a test code to test whether username_field is applied to the add_form.html template.
```
class UserAdminTest(TestCase):
    ...
    def test_guideline_username(self):
        username = User._meta.get_field(User.USERNAME_FIELD).verbose_name
        response = self.client.get(reverse("admin:auth_user_add"))
        self.assertContains(
            response,
            "<p>First, fill the fields: %s and password. "
            "Then, you’ll be able to edit more user options.</p>" % (username),
            html=True,
        )
        response = self.client.get(
            reverse("admin:auth_user_add") + "?%s=1" % IS_POPUP_VAR
        )
        self.assertContains(
            response,
            "<p>Fill the fields: %s and password.</p>" % (username),
            html=True,
        )
    ...
```
Test the input guideline messages in the add_form.html template when popup is applied and not.
If there is anything lacking about the test code, I would appreciate it if you could advise me.
I'd appreciate it if you could advise me on the test function name as well.
It was the best thing for my naming sense..

File: django/contrib/admin/templates/admin/auth/user/add_form.html [hunk 2]
```diff
@@ -3,9 +3,9 @@
 
 {% block form_top %}
   {% if not is_popup %}
-    <p>{% translate 'First, enter a username and password. Then, you’ll be able to edit more user options.' %}</p>
+    <p>{% blocktranslate %}First, fill the fields: {{ username }} and password. Then, you’ll be able to edit more user options.{% endblocktranslate %}</p>
```
Comment (sarahboyce) [inline django/contrib/admin/templates/admin/auth/user/add_form.html:6]:
```suggestion
    <p>{% blocktranslate %}First, fill the {{ username }} and password fields. Then, you’ll be able to edit more user options.{% endblocktranslate %}</p>
```

File: django/contrib/admin/templates/admin/auth/user/add_form.html [hunk 3]
```diff
@@ -3,9 +3,9 @@
...
   {% else %}
-    <p>{% translate "Enter a username and password." %}</p>
+    <p>{% blocktranslate %}Fill the fields: {{ username }} and password.{% endblocktranslate %}</p>
```
Comment (sarahboyce) [inline django/contrib/admin/templates/admin/auth/user/add_form.html:8]:
```suggestion
    <p>{% blocktranslate %}Fill the {{ username }} and password fields.{% endblocktranslate %}</p>
```

File: django/contrib/admin/templates/admin/auth/user/add_form.html [hunk 4]
```diff
@@ -15,4 +15,4 @@
 {% block admin_change_form_document_ready %}
   {{ block.super }}
   <script src="{% static 'admin/js/unusable_password_field.js' %}" defer></script>
-{% endblock %}
+{% endblock %}
```
Comment (sarahboyce) [inline django/contrib/admin/templates/admin/auth/user/add_form.html:18]:
```suggestion
{% endblock %}

```
(nit) revert this change

File: docs/releases/5.2.txt [hunk 5]
```diff
@@ -41,6 +41,9 @@ Minor features
   :ref:`extrabody <extrabody>` for adding custom code before the closing
   ``</body>`` tag.
 
+* The ``admin/auth/user/add_form.html`` input guideline message is changed
+  with a value according to user model username_field.
+
```
Comment (sarahboyce) [inline docs/releases/5.2.txt:46]:
```suggestion
```
We only add release notes for new features rather than cleanups/bug fixes

File: tests/admin_views/tests.py [hunk 6]
```diff
@@ -7299,6 +7299,24 @@ def test_password_mismatch(self):
...
+            html=True,
+        )
+        response = self.client.get(
+            reverse("admin:auth_user_add") + "?%s=1" % IS_POPUP_VAR
+        )
+        self.assertContains(
+            response,
+            "<p>Fill the fields: %s and password.</p>" % (username),
+            html=True,
+        )
```
Comment (sarahboyce) [inline tests/admin_views/tests.py:7318]: I think we want to test an admin with a custom user, something like...
```diff
--- a/tests/auth_tests/test_views.py
+++ b/tests/auth_tests/test_views.py
@@ -8,6 +8,7 @@ from urllib.parse import quote, urljoin
 from django.apps import apps
 from django.conf import settings
 from django.contrib.admin.models import LogEntry
+from django.contrib.admin.options import IS_POPUP_VAR
 from django.contrib.auth import BACKEND_SESSION_KEY, REDIRECT_FIELD_NAME, SESSION_KEY
 from django.contrib.auth.forms import (
     AuthenticationForm,
@@ -1758,3 +1759,40 @@ class UUIDUserTests(TestCase):
         # constraint error, delete the entry before constraints are checked
         # after the test.
         row.delete()
+
+
+@override_settings(
+    AUTH_USER_MODEL="auth_tests.CustomUser",
+    ROOT_URLCONF="auth_tests.urls_custom_user_admin",
+)
+class CustomUserAdminTests(TestCase):
+    @classmethod
+    def setUpTestData(cls):
+        cls.superuser = CustomUser._default_manager.create_superuser(
+            email="test@example.com",
+            password="test",
+            date_of_birth=datetime.date(2006, 4, 25),
+        )
+
+    def setUp(self):
+        self.client.force_login(self.superuser)
+
+    def test_guideline_username(self):
+        response = self.client.get(
+            reverse("custom_user_admin:auth_tests_customuser_add")
+        )
+        self.assertContains(
+            response,
+            "<p>First, fill the email address and password fields. "
+            "Then, you’ll be able to edit more user options.</p>",
+            html=True,
+        )
+        response = self.client.get(
+            reverse("custom_user_admin:auth_tests_customuser_add")
+            + "?%s" % IS_POPUP_VAR
+        )
+        self.assertContains(
+            response,
+            "<p>Fill the email address and password fields.</p>",
+            html=True,
+        )
diff --git a/tests/auth_tests/urls_custom_user_admin.py b/tests/auth_tests/urls_custom_user_admin.py
index 1c7ce1eb42..33f051c467 100644
--- a/tests/auth_tests/urls_custom_user_admin.py
+++ b/tests/auth_tests/urls_custom_user_admin.py
@@ -7,6 +7,14 @@ site = admin.AdminSite(name="custom_user_admin")
 
 
 class CustomUserAdmin(UserAdmin):
+    fieldsets = ((None, {"fields": (get_user_model().USERNAME_FIELD, "password")}),)
+    add_fieldsets = (
+        (
+            None,
+            {"fields": (get_user_model().USERNAME_FIELD, "password1", "password2")},
+        ),
+    )
+
     def log_change(self, request, obj, message):
         # LogEntry.user column doesn't get altered to expect a UUID, so set an
```

**Review by sarahboyce — COMMENTED** [review]
Thank you @Antoliny0919 :star: this is really close, I have minor suggestions

**Comment by Antoliny0919** [comment]
I have reflected all of your suggestions, thank you so much @sarahboyce !

File: tests/admin_views/tests.py (hunk 6 above)
Comment (Antoliny0919) [reply tests/admin_views/tests.py]: Reflecting the test code you suggested, there is no problem with the test in my local environment, but errors occur in the test process by PR.

**Reverse for 'auth_tests_uuiduser_change' not found. 'auth_tests_uuiduser_change' is not a valid view function or pattern name.**

```
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-focal/database/mysql/label/focal-pr/python/python3.12/tests/auth_tests/test_views.py", line 1729, in test_admin_password_change
    user_change_url = reverse(
                      ^^^^^^^^
```

As above, there is a problem in finding the logic associated with url through the reverse function and path_name.

I don't know why, but I'm having a problem with a path that uses the same ROOT_URLCONF value, urls_custom_user_admin.

Do you know anything about how to solve this problem?
I will also look for more to resolve this error. 💪

Comment (sarahboyce) [reply tests/admin_views/tests.py]: Ah sorry, I pushed up a tweak. I think the test ordering and initialization of `get_user_model()` was causing a problem

Comment (Antoliny0919) [reply tests/admin_views/tests.py]: @sarahboyce Thank you for your help. 💪💪

File: django/contrib/admin/templates/admin/auth/user/add_form.html (hunk 1 above)
Comment (sarahboyce) [reply django/contrib/admin/templates/admin/auth/user/add_form.html]: I just tweaked this to be:

`First, enter a valid {{ username }} and password...` 

to be consistent to the existing translatable strings like 
- `Please enter a correct %(username)s and password. Note...`
- `Please enter the correct %(username)s and password for a staff account...`

However, @claudep was the point that we force translators to construct a dynamic sentence and a listing structure would be better for all of the translatable strings? (And so we should update all strings to be like `Fill the fields: %(username)s and password. Note...` and `Fill the staff account fields: %(username)s and password.`

File: django/contrib/admin/templates/admin/auth/user/add_form.html [hunk 7]
```diff
@@ -3,9 +3,9 @@
 
 {% block form_top %}
   {% if not is_popup %}
-    <p>{% translate 'First, enter a username and password. Then, you’ll be able to edit more user options.' %}</p>
+    <p>{% blocktranslate %}First, enter a valid {{ username }} and password. Then, you’ll be able to edit more user options.{% endblocktranslate %}</p>
```
Comment (cliff688) [inline django/contrib/admin/templates/admin/auth/user/add_form.html:6]: I'm not sure about "valid", since this is a user creation form. If a field is invalid I think we can let them know in the errors.

```suggestion 
    <p>{% blocktranslate %}First, enter the {{ username }} and password. Then, you’ll be able to edit more user options.{% endblocktranslate %}</p>
```
Maybe the definite particle might be safe here: as in "*the* username and password for the new user." I don't know how that would translate in other languages, though.

Comment (sarahboyce) [reply django/contrib/admin/templates/admin/auth/user/add_form.html:6]: I agree using "the" is nicer than my suggestion :+1:

**Comment by Antoliny0919** [comment]
So far, I have found two parts where username has been used.
```
# admin
"Please enter the correct %(username)s and password for a staff account. Note "
"that both fields may be case-sensitive."

# auth
"Please enter a correct %(username)s and password. Note that both fields may "
"be case-sensitive."
```
With the above two and the review given by @claudep, I think the following three are candidates to be guidelines messages.
```
1. First, enter a {{ username }} and password.

2. First, enter the {{ username }} and password.

3. First, fill the fields: {{ username }} and password.
```
Considering the consistency, it's number 1 and 2, but I think number 3 is also a good choice considering the translation according to the review claudep gave me.

I'm curious about the opinions of members and contributers. What should I use as a guideline message?

File: django/contrib/admin/templates/admin/auth/user/add_form.html (hunk 1 above)
Comment (Antoliny0919) [reply django/contrib/admin/templates/admin/auth/user/add_form.html]: I looked at two parts where username was used similarly.
```
# admin
"Please enter the correct %(username)s and password for a staff account. Note "
"that both fields may be case-sensitive."

# auth
"Please enter a correct %(username)s and password. Note that both fields may "
"be case-sensitive."
```
Is there a problem if I modify the message part of add.html similar to the above?
```
First, enter (a or the) {{ username }} and password.
```
Can you check this one more time please?

File: django/contrib/admin/templates/admin/auth/user/add_form.html (hunk 7 above)
Comment (claudep) [reply django/contrib/admin/templates/admin/auth/user/add_form.html:6]: Unfortunately, this string is still not translator-friendly. First, the `username` variable name is misleading, as it can contains anything else, and then in many languages, the `username` placeholder will need some declination due to its grammatical position. Are we forced to name the fields in this sentence, couldn't we mention something like `the following fields` or some equivalent formulation that doesn't require the field names?

Comment (Antoliny0919) [reply django/contrib/admin/templates/admin/auth/user/add_form.html:6]: I think the current sentence is the best if we keep the static form.
However, I think the sentence feels awkward from the users point of view if the username field changes like the issue we are dealing with now.
So I think it's better to use dynamic values even if we consider translation.
How about the sentence that @claudep suggested?
```
First, fill the fields: {{ username }} and password.
```
If there is no problem with the translation, I think this sentence is also good.

**Comment by Antoliny0919** [comment]
I confirmed that it becomes more awkward when I used USERNAME_FIELD only for add_form.html.
As an example, if I change USERNAME_FIELD to email and add_form.html to use the USERNAME_FIELD value, it is as follows.
Screenshot 2024-09-26 at 11 05 09 AM
As above, the changed input guidelines show an email, but the form still has a username.
Therefore, to configure dynamic values through USERNAME_FIELD, we must also use USERNAME_FIELD in UserAdmin's add_fieldset.
```
@admin.register(User)
class UserAdmin(admin.ModelAdmin):
   ...
    add_fieldsets = (
        (
            None,
            {
                "classes": ("wide",),
                "fields": (User.USERNAME_FIELD, "usable_password", "password1", "password2"),
            },
        ),
    )
    ...
```
I thought this change was an improvement in UserAdmin provided by Django, so I modified the related [issue](https://code.djangoproject.com/ticket/35742).

**Review by sarahboyce — COMMENTED** [review]
Thank you for your work on this @Antoliny0919

I have had a think and I think it is best to update the messages to avoid mentioning specific fields like: `After you've created a user, you’ll be able to edit more user options.`

We also have Natalia's [ticket comment](https://code.djangoproject.com/ticket/35742#comment:11) which needs to be addressed: 
- `Forgotten your password or username?` -> maybe `Forgotten your login credentials?`
- `Your username, in case you’ve forgotten:` -> maybe `Your other credentials, in case you’ve forgotten:`

With the simpler solution, the tests can be simplified to checking normal admin views rather than a custom user admin

I would leave the `UserAdmin` alone as this is designed for models inheriting from `AbstractUser` rather than `AbstractBaseUser` and so they will have a username

**Comment by Antoliny0919** [comment]
I applied the parts you suggested. @sarahboyce 
And I also applied the parts suggested by @nessita on the ticket.

I have one question.
There are some problems that users may experience when using USERNAME_FIELD for UserAdmin in the way I suggested.

Assuming that the user changes USERNAME_FIELD to a different field using a custom user model, the most likely candidate is "email."

If the user changes USERNAME_FIELD to one of the fields provided by Django AbstractUser, the fieldset value of UserAdmin must also be changed.
Because fieldsets return errors for duplicate field values.
For example, when I changed USERNAME_FIELD to "email" and applied UserAdmin.
Screenshot 2024-10-08 at 10 09 05 AM
The above error occurs because the fieldsets do not allow duplicate fields, as I have previously explained.
```
username = get_user_model().USERNAME_FIELD # --> email

@admin.register(User)
class UserAdmin(admin.ModelAdmin):
    add_form_template = "admin/auth/user/add_form.html"
    change_user_password_template = None
    fieldsets = (
        (None, {"fields": (username --> "email", "password")}),
        (_("Personal info"), {"fields": ("first_name", "last_name", "email")}), 
        ...
```
The list_display in ModelAdmin is also affected and displays duplicate fields.
Screenshot 2024-10-08 at 10 13 28 AM
Of course, removing duplicates of those fields will fix the problem.

However, I thought these parts might rather inconvenience Django users, so I was curious about other people's opinions.

**Comment by sarahboyce** [comment]
> Assuming that the user changes USERNAME_FIELD to a different field using a custom user model, the most likely candidate is "email."

Agreed, although email needs to be overridden to be a unique field and won't work out of the box. 
The other option is a new field

> If the user changes USERNAME_FIELD to one of the fields provided by Django AbstractUser, the fieldset value of UserAdmin must also be changed.

I don't think there is a perfect solution given that the user has to be overriding the model.
The minimal change is doing just the templates, as those are the hardest to override.

I wouldn't worry about the list and search fields - they still "work"

The main thing for me is can a User be created and updated? I think the rest of the work should be done as part of ticket-28608 in a separate PR

**Comment by Antoliny0919** [comment]
> I don't think there is a perfect solution given that the user has to be overriding the model.
The minimal change is doing just the templates, as those are the hardest to override.

Yes, I agree with you.
If so, is there any more work I have left in this PR?
Please let me know if there is any more work left and I will proceed.

**Review by sarahboyce — APPROVED** [review]
Thank you @Antoliny0919 :star: and welcome onboard!
Plan to land this some point today
